File: 01-Algorithmic-toolbox/python/01-max_pairwise_product.py
# Uses python3
n = int(input("Enter n "))
a = [int(x) for x in input("Enter numbers ").split()]
assert(len(a) == n)

result = 0
# Find the first and second maximum number in array
if n <= 1:
    print(result)
else:
    max1 = a[0] if a[0] > a[1] else a[1]
    max2 = a[1] if a[0] > a[1] else a[0]
    for i in range(2, n):
        if 	a[i] > max1:
            max1, max2 = a[i], max1
        elif a[i] > max2 and a[i] <= max1:
            max2 = a[i]

# the result is the multiplication of these numbers
result = max1*max2

print(result)

File: 01-Algorithmic-toolbox/javascript/01-max_pairwise_product.js
function program() {
    console.log("Welcome to Max Pairwise program");
    const readline = require("readline").createInterface({
        input: process.stdin,
        output: process.stdout
    })

    let n = 0;
    let numbers = [];
    readline.question("Enter n: ", number => {
        n = parseInt(number);
        readline.question("Enter numbers ", input => {
            numbers = input.split(' ').map(Number);
            readline.close();

            console.log(calculateMaxPairwise(n, numbers));
        });
    })
}

function calculateMaxPairwise(n, numbers) {
    if (n < 1) {
        return -1;
    } else if (n == 1)
    {
        return n;
    } else if (n == 2)
    {
        return numbers[0] * numbers[1];
    }

    let max1 = numbers[0] > numbers[1] ? numbers[0] : numbers[1];
    let max2 = numbers[0] > numbers[1] ? numbers[1] : numbers[0];

    for (let i = 2; i <= numbers.length; i++) {
        if (numbers[i] > max1) {
            max2 = max1;
            max1 = numbers[i];
        } else if (numbers[i] > max2) {
            max2 = numbers[i]
        }
    }

    return max1 * max2;
}

program();

